File: nodejs/src/Sort/GroupAnagrams/GroupAnagrams.ts
class GroupAnagrams {
  strs: string[]

  constructor(strs: string[]) {
    this.strs = strs
  }

  /**
   * we will use sorted string as key
   */
  exec(): string[][] {
    const h: Map<string, string[]> = new Map<string, string[]>();

    for(let str of this.strs){
      const key: string = str.split('').sort().join('')
      const data: string[] = h.get(key) || []
      data.push(str)
      h.set(key, data)
    }

    return [...h.values()]
  }
}

export default GroupAnagrams;

File: python/src/hash/group_anagrams.py
from typing import List
from collections import defaultdict


class Solution:
    def groupAnagrams(self, strs: List[str]) -> List[List[str]]:
        count_to_values = defaultdict(list)
        for s in strs:
            # think of using array for string problem with character set limited
            # to lowercase English letters
            count = [0] * 26
            for c in s:
                count[ord(c) - ord("a")] += 1
            count_to_values[tuple(count)].append(s)

        return list(count_to_values.values())

    def groupAnagrams2(self, strs: List[str]) -> List[List[str]]:
        str_to_values = defaultdict(list)

        for s in strs:
            key = "".join(sorted(s))
            str_to_values[key].append(s)

        return list(str_to_values.values())
